aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -42,6 +42,23 @@ static inline DispatchKey dispatchTypeId(
     DispatchKeySet key_mask
 ) {
   c10::impl::LocalDispatchKeySet local = c10::impl::tls_local_dispatch_key_set();
+  // At most one Autograd backend key is considered for computing dispatch key.
+  // Imagine an op takes both CPU tensor and PrivateUse1 tensor as inputs, we have
+  // - a kernel k1 registered as catchAll
+  // - a kernel k2 registered to PrivateUse1
+  // The computed dispatch table entry looks like:
+  //   CPU: k1 [catchAll]
+  //   PrivateUse1: k2 [kernel]
+  //   AutogradCPU: k1 [catchAll]
+  //   AutogradPrivateUse1: fallthrough [backend fallback]
+  // In this case it calls dispatcher with
+  //   ks = {AutogradPrivateUse1, AutogradCPU, PrivateUse1, CPU}
+  // After we fall through AutogradPrivateUse1 key, we want to dispatch to backend
+  // key PrivateUse1 instead of AutogradCPU. To do that, we make ks contain at most
+  // one Autograd dispatch key.
+  DispatchKeySet autograd_keys = getRuntimeDispatchKeySet(c10::DispatchKey::Autograd);
+  DispatchKey highest_autograd = (ks & autograd_keys).highestPriorityTypeId();
+  ks = (ks - autograd_keys).add(highest_autograd);
```

Comment:
This looks very bad. We should work hard to avoid adding all of this extra work in the runtime dispatch path.

Reply:
If we can guarantee that (using Ailing's comment example) an autograd key of `AutogradPrivateUse1` will fall through to `PrivateUse1`, then we can precompute I think.

Reply:
In VC, we established that this would not work, as AutogradOther is a one to many mapping. (I also argued that if we ever added more keys between backend and autograd, this would also not work).

Reply:
Agreed, precomputation wouldn't work for `AutogradOther` (which we decided should be used instead of `AutogradPrivateUse1` in this case, for reasons outside the scope of this PR) bc it's one-to-many. 

We also decided that the desired end state is not to use/allow fallthrough kernels on autograd keys at all, so the following no longer bears directly on this PR, but just to be a little more precise about the conditions where fallthrough precomputation would work so it doesn't go down the memory hole:
* one-to-many mappings _can't_ work (since they break the guarantee mentioned in the earlier comment, that a fallthrough's destination kernel be uniquely determined). One-to-one and many-to-one mappings can.
    * you could actually generalize this to only require that the "final destination" of a fallthrough needs to be uniquely determined - precomputation could fall through any number of non-unique intermediate stages, as long as they were populated entirely by fallthroughs, and all paths terminated at the same non-fallthrough kernel.
* new intermediate stages (e.g. "more keys between backend and autograd") that preserve the first condition _can_ work, but would need an upgrade to the precomputation algorithm to teach it the new keys.

Reply:
Btw created https://github.com/pytorch/pytorch/issues/43908 to track that issue. IIRC you mentioned there're some boxed/unboxed stuff blocking us moving to the redispatch as fallback Autograd kernel? Would you mind updating that thread with some context? ;) Thanks a lot!